datafusion/core/src/physical_plan/unnest.rs

```diff
@@ -264,41 +267,224 @@ fn unnest_batch<T>(
+/// the final batch:
+///
+/// ```ignore
+/// c1: 1, null, 2, 3, 4, null, 5, 6
+/// ```
+///
+/// And the `indices` array contains the indices that are used by `take` kernel to
+/// repeat the values in `c2`:
+///
+/// ```ignore
+/// 0, 1, 2, 2, 2, 3, 4, 4
+/// ```
+///
+/// so that the final batch will look like:
+///
+/// ```ignore
+/// c1: 1, null, 2, 3, 4, null, 5, 6
+/// c2: 'a', 'b', 'c', 'c', 'c', null, 'd', 'd'
+/// ```
+///
+fn batch_from_indices<T>(
+    batch: &RecordBatch,
+    schema: &SchemaRef,
+    unnest_column_idx: usize,
+    unnested_array: &ArrayRef,
+    indices: &PrimitiveArray<T>,
+) -> Result<RecordBatch>
+where
+    T: ArrowPrimitiveType,
+{
+    let arrays = batch
+        .columns()
+        .iter()
+        .enumerate()
+        .map(|(col_idx, arr)| {
+            if col_idx == unnest_column_idx {
+                Ok(unnested_array.clone())
+            } else {
+                Ok(kernels::take::take(&arr, indices, None)?)
+            }
+        })
+        .collect::<Result<Vec<_>>>()?;
+
+    Ok(RecordBatch::try_new(schema.clone(), arrays.to_vec())?)
+}
+
+/// Unnest the given list array. Given the array:
+///
+/// ```ignore
+/// [1], null, [2, 3, 4], null, [5, 6]
+/// ```
+///
+/// returns:
+///
+/// ```ignore
+/// 1, null, 2, 3, 4, null, 5, 6
+/// ```
+fn unnest_array<T>(list_array: &T) -> Result<Arc<dyn Array + 'static>>
+where
+    T: ArrayAccessor<Item = ArrayRef>,
+{
+    let elem_type = match list_array.data_type() {
+        DataType::List(f) | DataType::FixedSizeList(f, _) | DataType::LargeList(f) => {
+            f.data_type()
+        }
+        dt => {
+            return Err(DataFusionError::Execution(format!(
+                "Cannot unnest array of type {dt}"
+            )))
+        }
+    };
+
+    let null_row = new_null_array(elem_type, 1);
+
+    // Create a vec of ArrayRef from the list elements.
+    let arrays = (0..list_array.len())
+        .map(|row| {
+            if list_array.is_null(row) {
+                null_row.clone()
+            } else {
+                list_array.value(row)
+            }
+        })
+        .collect::<Vec<_>>();
+
+    // Create Vec<&dyn Array> from Vec<Arc<dyn Array>> for `concat`. Calling
+    // `as_ref()` in the `map` above causes the borrow checker to complain.
+    let arrays = arrays.iter().map(|a| a.as_ref()).collect::<Vec<_>>();
+
+    Ok(kernels::concat::concat(&arrays)?)
+}
+
+/// Returns an array with the lengths of each list in `list_array`. Returns null
+/// for a null value.
+fn list_lengths<T>(list_array: &T) -> Result<Arc<dyn Array + 'static>>
+where
+    T: ArrayAccessor<Item = ArrayRef>,
+{
+    match list_array.data_type() {
+        DataType::List(_) | DataType::LargeList(_) => {
+            Ok(kernels::length::length(list_array)?)
+        }
+        DataType::FixedSizeList(_, size) => {
+            // Handle FixedSizeList as it is not handled by the `length` kernel.
```

Comment:
Filed https://github.com/apache/arrow-rs/issues/4517 to trach this

```suggestion
            // Handle FixedSizeList as it is not handled by the `length` kernel.
            // https://github.com/apache/arrow-rs/issues/4517
```